### ISSUE TITLE
Question on Partial Weight Selection in Prefill Stage

### ISSUE THREAD

**Issue by Serenii02** [snu-comparch/InfiniGen#7]
Hi authors,

Thank you for your great work on InfiniGen!

While reading Section 4.3 of your paper, I noticed the following description in the prefill stage:

"To obtain partial matrices that capture the outliers, we first take the element-wise absolute values of the skewed query and key matrices, then **add these two matrices together.** This helps us calculate the sum of each column and perform top-k operation only once while accommodating the outlier columns of both query and key matrices."

From this description, my understanding is that we should first compute the element-wise absolute values of the skewed query and key matrices, add them together, and then perform a single top-k operation on the summed matrix to select the important columns.

However, when I checked the implementation in _InfiniGen/accuracy/src/modeling_llama_ours_setup.py_ (lines 214–224), it seems that the code does the following instead:
```python
if self.partial_weight_q is None:
    weight_mask = torch.zeros_like(self.q_proj.weight.data.t())
    # Speculate major columns in Wq and Wk
    for head in range(int(query.shape[-1] / self.head_dim)):
        start = head * self.head_dim
        end = (head+1) * self.head_dim
        _, ind = torch.topk(torch.sum(torch.abs(query_states[0, head] @ self.skewing_matrix[head]), dim = -2), int(self.head_dim * self.partial_weight_ratio))
        weight_mask[:, start:end] = weight_mask[:, start:end].scatter(-1, ind.repeat(weight_mask.shape[0], 1), 1)
        _, ind = torch.topk(torch.sum(torch.abs(key_states[0, head] @ self.skewing_matrix[head]), dim = -2), int(self.head_dim * self.partial_weight_ratio))
        weight_mask[:, start:end] = weight_mask[:, start:end].scatter(-1, ind.repeat(weight_mask.shape[0], 1), 1)
    self.partial_weight_q = weight_mask
```

It looks like in the code, **top-k is applied separately on the absolute value matrices of the query and key**, and the corresponding indices are updated sequentially. I feel that this implementation might not be strictly equivalent to the description in the paper (which suggests combining them first before top-k).

I’m not sure if I’ve understood this part correctly. Could you kindly clarify whether my understanding is accurate?

Thank you very much for your time and help!

**Comment by leewonbeom** [comment]
Hello, 

Thank you for catching this discrepancy.
There is a small inconsistency between the paper description and the implementation.

This arose from overlapped paper submission and artifact evaluation timelines. 
We refined the approach in the paper, but did not update the code since the artifact evaluation was already underway.

We will fix this inconsistency in the next update by making the code follow the description in the paper. 
Thanks for the careful review!
